tools/single_use/remove_logged_passwords.php

```diff
@@ -0,0 +1,47 @@
+<?php
+require_once __DIR__ . "/../generic_includes.php";
+$info = <<<INFO
+This script deletes occurrences of plaintext passwords that were stored in the
+`history` table.
+
+This script will also expire all current passwords for LORIS users. 
+
+The plaintext passwords appeared due to a bug in LORIS. As these passwords
+have potentially been viewed by administrators or could be read in the event
+of a data breach, we have chosen to reject them in this way.
```

Comment:
The "expire all passwords" should be an option (or disabling it should be an option, at least) toggled by a flag to the script.

Reply:
I'm reluctant to hamstring this fix. Since the passwords are available in plaintext in the back-end I think they should be considered unfit to use in the future. This was the original idea behind implementing a blacklist. While the fix no longer includes the blacklisting functionality, I feel that forcing a password reset is a minimum requirement to properly address this issue.

Reply:
Letting projects decide their own policy rather than forcing yours on them isn't "hamstringing the fix". The only people who potentially could read from the history table/audit trail (which isn't exposed anywhere) can already trivially change a user's password* if they wanted to access their account. (In fact, given that they have to already have backend access, they could even change the password, login as the user, then update the password hash back to the original in the user table to do it without a trace, with or without this fix.) Expiring the password in LORIS does not add any extra security, as the risk is password re-use on other sites that you don't have any ability to expire passwords on. Choosing to do so or not should be up to the projects using LORIS to decide.

* There are also scripts and frontends where an admin could reset a user's password to a known value that are even less restrictive than update access.

Reply:
This, I feel, is the responsible path for dealing with the issue of an internal password exposure:
* Delete the logged plaintext passwords
* Fix the code so that this doesn't happen again
* Notify the users that their password have been revealed in the database
* Reset the logged passwords
* Strongly recommend that the users reset passwords on other sites where they use the same password (if any -- but it's probable)

I think the level of risk is greater than you're making it out to be. You are describing what a current project administrator can do but this is not the only threat scenario. These passwords would also have been visible to people that no longer work at the facilities using LORIS who no longer have any other means of accessing the software. There is also the real possibility that the database could have been hacked into... even now we have a known SQL injection vector so it's not an imagined risk.

As you said a primary issue here is password reuse and of course we can't control what users do on other sites. But we can encourage best practices within the LORIS domain by triggering a password reset along with an explanation of what occurred. This puts the potential of exposure in the user's mind and if they are made to change their LORIS password because it has potentially been revealed, they may also reset their passwords elsewhere if they are prudent. So we can't reset passwords elsewhere but we can suggest that users do so of their own choice. A quieter patch without a reset makes that moment of reflection less likely which means a greater risk to the people involved.

We should respond to this bug in the best way possible and a password reset + explanation is the best way. As always a project administrator is able to edit or ignore the patch; nobody is being forced to do anything.

Reply:
There is no proof of internal password exposure. There is a theoretical possibility of it because it was logged in the most restrictive place which is never intentionally exposed anywhere. If you have any evidence of someone having illegally accessed it from there with the intent of stealing user's passwords, you should be filling a criminal report, not a pull request.

Even though it's probably impossible to tell if someone *did* access it, your script precludes the possibility of notifying users who *may* have had someone read it, since it deletes entries in the history table without having logged who they were first and just blindly expires everyone's password.

If you don't want to add a flag to to allow projects to decide their own policies on passwords expiry without having to modify code you should (in fact, it's probably a better idea than having a flag now that I think about it) be printing a report of "UserID Email Active Y/N (Date(s) Of Change)" before deleting them so that they can be contacted. After having a query that does that, those are the only users who are at any risk and should have their password expired.